### ISSUE TITLE
PHP / Psalm installed in docker container

### ISSUE THREAD

**Issue by shaneparsons** [psalm/psalm-vscode-plugin#176]
I have php / psalm running in a docker container... I've tried various different config options trying to get this plugin to find psalm in that container, but there doesn't seem to be any support for such a workflow... Any ideas / tips? Anybody else have success with something similar?

**Comment by weirdan** [comment]
Psalm (including the language server) does work in container, however there's currently no support for mapping the paths between host machine and the container. There's a PR in the Psalm repo to address this: vimeo/psalm#8026

**Comment by tm1000** [comment]
My team uses docker all day long. We actually execute and run psalm outside of the docker instance (so we use our local php binaries). This works well in our case and has been that way for the last 2 years we've been using psalm. We have about 4-5 developers working this way and don't experience issues.

The way the plugin itself works is through STDIO pipes but you can also configure it to connect through TCP however it binds to localhost and is assigned a random port. In TCP Mode vscode starts a language server and then tells Psalm to start with the `--tcp` option which tells Psalm to connect to said ip and port. If you setup docker correctly you could route the connection outward (so vscode starts the process then you tell psalm to connect to said process with the `--tcp` option). You'd also need to declare `extra_hosts`

EG:
```
    extra_hosts:
        - 'host.docker.internal:host-gateway'
```

With this set inside the docker container itself I am able to ping the host. 

```
root@e50b0fd4c96e:/usr/local/app# ping host.docker.internal
PING host.docker.internal (172.17.0.1) 56(84) bytes of data.
64 bytes from host.docker.internal (172.17.0.1): icmp_seq=1 ttl=64 time=1.48 ms
64 bytes from host.docker.internal (172.17.0.1): icmp_seq=2 ttl=64 time=0.060 ms
64 bytes from host.docker.internal (172.17.0.1): icmp_seq=3 ttl=64 time=0.099 ms
64 bytes from host.docker.internal (172.17.0.1): icmp_seq=4 ttl=64 time=0.067 ms
```

This is how we get xdebug breaks to work.

Let me know if this is something you'd like as looking at the code it doesn't seem hard to achieve. We'd basically expose the ability to define the port and ip for both the server and the client ahead of time. In the ip for the client you'd set it to `host.docker.internal` which would turn into `--tcp=host.docker.internal:<someport>`

**Comment by shaneparsons** [comment]
Hey @tm1000, sorry for the delay. That sounds great though! similar to what https://github.com/plotbox-io/psalm-docker-vscode is doing actually.

In working to get the other plugin working in vscode I actually ran into some issues with the container not being able to see the outside, even though I was able to ping it just fine. I think it had to do with the port being randomized, so the ability to set the port would be great. You can actually read a lot about my setup / experience in [this issue](https://github.com/plotbox-io/psalm-docker-vscode/issues/1), if interested.

Currently my workflow is that I moved one of the smaller projects I work on locally, so that php exists on the system and I can use this plugin as is. But there are other bigger / older projects with different php versions that I can't easily do this with, so being able to use the containerized php would be awesome.

**Comment by weirdan** [comment]
> however there's currently no support for mapping the paths between host machine and the container.

There is now: https://github.com/vimeo/psalm/pull/10033
This extension needs to allow specifying `--map-folder` flag for this to work in vscode though.

**Comment by kgilden** [comment]
> Let me know if this is something you'd like as looking at the code it doesn't seem hard to achieve. We'd basically expose the ability to define the port and ip for both the server and the client ahead of time. In the ip for the client you'd set it to `host.docker.internal` which would turn into `--tcp=host.docker.internal:<someport>`

Could you clarify how exactly the ip for the client can be configured? It seems that vscode starts the TCP server at 127.0.0.1 and  so even though the host is reachable from within the container via host.docker.internal, the TCP server is refusing connections.

Reading though the code of the extension it seems to me that 127.0.0.1 is hard-coded.

```js
server.listen(0, '127.0.0.1', () => {
    // Start the language server
    // make the language server connect to the client listening on <addr> (e.g. 127.0.0.1:<port>)
    this.spawnServer([
        // @ts-ignore
        '--tcp=127.0.0.1:' + server.address().port,
    ]);
});
```

Or are you talking about a potential solution *that needs to be built* to run psalm inside a container?

----

Why am I trying to get TCP mode to work? I can''t seem to get vscode talking with the psalm language server using STDIN. The language server starts up fine in a container and path names can be mapped just fine. Yet for some unknown reason the messages sent by vscode never seem to reach the language server via STDIN.

**Comment by tm1000** [comment]
@kgilden I dont think this was looked into deep enough to be honest. At least in my use case Psalm is run outside of the docker container in the host and works fine though of course the host has to be equal to or greater than what psalm wants in the container. If those match then paslm itself will read the composer.json file to determine its php feature set. So for my development team I keep them on the highest PHP version (8.4 soon to be 8.5)

**Comment by kgilden** [comment]
Thanks. Yep, it takes some hacking to get TCP mode working in the current version of the extension.

For better DX I would propose the following changes to this project.

1. Make the host server listening address configurable, don't keep it hard-coded at 127.0.0.1. Not sure if necessary, but perhaps also let users configure the address to connect (in the example below I had to use host.docker.internal).
2. Add support for variables in config so that I could for example do `"psalm.phpExecutablePath": "${workspaceFolder}/bin/myscript.sh",`. Right now it's impossible to make the config portable across different machines.

----

Here's a list of hoops I had to jump through. Note that this is very ugly and is so not portable.

1. Hack the extension to make the server on the the host start listening to everything (note, this is insecure, can this be improved?).
   ```
   # Replace 127.0.0.1 with 0.0.0.0 inside these files depending if using oss or non-oss vscode.
   ~/.vscode-oss/extensions/getpsalm.psalm-vscode-plugin-x.y.x/dist/extension.js
   ~/.vscode/extensions/getpsalm.psalm-vscode-plugin-x.y.x/dist/extension.js
   ```
2. Make the host machine accessible from within the container. When using `docker compose`, ensure the following config is added to your container.
   ```yaml
   extra_hosts:
     - "host.docker.internal:host-gateway"
   ```
3. Ensure the following settings have been configured in settings.json.
   ```json5
   {
     // ...
     "psalm.phpExecutablePath": "/path/to/project/on/host/.vscode/php-wrapper.sh",
     "psalm.psalmScriptPath": "vendor/bin/psalm.phar",
     "psalm.psalmScriptArgs": ["--map-folder=/path/to/project/on/host:/path/to/project/in/container"],
     "psalm.connectToServerWithTcp": true,
     // ...
   }
   ```
4. Add the following script to `/path/to/project/on/host/.vscode/php-wrapper.sh`.
   ```sh
   #!/bin/bash
   # Wrapper script to run PHP from the container for VS Code extensions
   
   WORKSPACE_DIR="$(cd "$(dirname "$0")/.." && pwd)"
   CONTAINER_ROOT="/path/to/project/in/container"
   
   # Translate host paths to container paths
   args=()
   for arg in "$@"; do
       # Don't translate --map-folder arguments (they need host paths)
       if [[ "$arg" == --map-folder=* ]]; then
           args+=("$arg")
       # Translate --tcp=127.0.0.1:PORT to --tcp=host.docker.internal:PORT
       # so the language server inside the container can reach VS Code on the host
       elif [[ "$arg" == --tcp=127.0.0.1:* ]]; then
           port="${arg#--tcp=127.0.0.1:}"
           args+=("--tcp=host.docker.internal:${port}")
       else
           # Replace workspace paths with container root
           translated="${translated//${WORKSPACE_DIR}/${CONTAINER_ROOT}}"
           args+=("$translated")
       fi
   done
   
   cd "$WORKSPACE_DIR" || exit 1
   docker-compose exec my-container-name php "${args[@]}"
   ```
